src/rgw/rgw_data_sync.cc

```diff
@@ -2035,6 +2085,196 @@ int RGWReadBucketSyncStatusCoroutine::operate()
   }
   return 0;
 }
+
+class RGWReadBucketShardsStatusCR : public RGWShardCollectCR {
+  static constexpr int MAX_CONCURRENT_SHARDS = 16;
+
+  RGWDataSyncEnv *sync_env;
+
+  map<rgw_bucket_shard, rgw_bucket_shard_sync_info> &bs_sync_status;
+  map<rgw_bucket_shard, rgw_bucket_shard_sync_info>::iterator iter;
+  
+ public:
+  RGWReadBucketShardsStatusCR(RGWDataSyncEnv *_sync_env, 
+      map<rgw_bucket_shard, rgw_bucket_shard_sync_info>& _bs_sync_status)
+    : RGWShardCollectCR(_sync_env->cct, MAX_CONCURRENT_SHARDS), sync_env(_sync_env),
+    bs_sync_status(_bs_sync_status)
+  {
+    iter = bs_sync_status.begin();
+  }
+  bool spawn_next() override;
+};
+
+bool RGWReadBucketShardsStatusCR::spawn_next()
+{
+ if (iter == bs_sync_status.end()) {
+   return false;
+ }
+
+ rgw_bucket_shard bs = iter->first;
+ spawn(new RGWReadBucketSyncStatusCoroutine(sync_env, bs, &iter->second), false);
+
+  ++iter;
+  return true;
+}
+
+#define DATA_SYNC_MAX_ACTIVE_ENTRIES 10
+class RGWReadLaggingBucketShardsCoroutine : public RGWCoroutine {
+  RGWDataSyncEnv *sync_env;
+  RGWRados *store;
+  
+  map<string, bucket_instance_meta_info> instance_entries;
+  set<string>& lagging_buckets;
+  map<rgw_bucket_shard, rgw_bucket_shard_sync_info> bs_sync_status;
+
+  const int shard_id;
+  uint64_t max_entries;
+  string marker;
+  set<string> entries;
+
+  string status_oid;
+
+  int id;
+  rgw_bucket_shard_sync_info sync_info;
+  rgw_bucket_shard bs;
+
+  rgw_data_sync_marker* sync_marker;
+public:
+  RGWReadLaggingBucketShardsCoroutine(RGWDataSyncEnv *_sync_env, const int _shard_id,
+                                      set<string>& _lagging_buckets, rgw_data_sync_marker* _sync_marker) 
+  : RGWCoroutine(_sync_env->cct), sync_env(_sync_env),
+  store(sync_env->store), lagging_buckets(_lagging_buckets),
+  shard_id(_shard_id), max_entries(DATA_SYNC_MAX_ACTIVE_ENTRIES),
+  sync_marker(_sync_marker)
+  {
+    status_oid = RGWDataSyncStatusManager::shard_obj_name(sync_env->source_zone, shard_id);
+  }
+
+  int operate() override;
+};
+
+int RGWReadLaggingBucketShardsCoroutine::operate()
```

Comment:
i worry that this algorithm will get too expensive in systems with a lot of buckets/shards. we're trying to provide a summary of pending work on a single datalog shard, but this approach requires gathering a list of all bucket instances in the realm - and we'd have to repeat that work for every datalog shard we want to inspect with this command

consider instead that we simply read entries from the local datalog shard, starting at our sync status marker, and print the next ~10 unique bucket shards as a list of 'pending bucket shards'. i would argue that this list is actually more useful to admins than an exhaustive list of all possible bucket shards because if data sync is behind, it's the bucket shards in this window that are the most likely to be the cause

in addition to the list of 'pending bucket shards', the list of bucket shards in the error repo ('recovering bucket shards') would also be useful to list here

maybe add an optional --max-entries option, and list that many entries from both the 'pending bucket shards' and 'recovering bucket shards' with a '...' at the end if there are more entries after that

what do you think?

Reply:
@cbodley List all bucket instance to get the bucket shards related with the specified data log shard, so that we can avoid reading all of the remote datalog entries and get all lagging bucket shards , that's my original purpose.
 Add some option to limit the number of entries sounds more reasonable,  but I'm afraid it may get too expensive too if the datalog shard have too many continuous entries  related with some bucket shards.
In addition, if --max-entries is somehow too large,  it will take too long to get the result.

Reply:
i don't think we need to list the remote datalog for this; during incremental sync, we pull 1000 remote datalog entries at a time into our local copy - that should be plenty to generate this list

in fact, this part is almost identical to the 'datalog list' command, except that we're filling in the --start-time based on sync status and filtering out duplicate entries